File: data-structures/stack/python/stack.py
"""
Stack data structure implementation based on linked lists
"""

# Author: 
# Lajos Neto <lajosnetogit@gmail.com>


class _Node:
    """
    Simple node used for representing values 
    inside the stack.
    """
    def __init__(self, value):
        self.value = value
        self.next = None
    
class Stack:
    """
    Stack data structure.
    """
    def __init__(self):
        self._top = None
        self._bottom = None
        self._size = 0
    
    def empty(self):
        """Check if the list is empty"""
        return self._size == 0

File: data-structures/stack/kotlin/Stack.kt
/**
 * Stack implementation.
 *
 * Author : Lajos Neto.
 * Email : lajosnetogit@gmail.com
 */

data class Node<T>(
    val value:T,
    var next: Node<T>? = null
)

class Stack<T>{
    private var size = 0
    private var head: Node<T>? = null

    override fun toString(): String {
        var stringRep = ""
        var walker = head
        while (walker != null) {
            stringRep += "${walker.value} -> "
            walker = walker.next
        }
        return "${stringRep}/null"
    }

    fun push(value: T) {
        val newNode = Node(value)
        newNode.next = head
        head = newNode
        size += 1
    }

    fun pop() {
        if (!isEmpty()) {
            head = head?.next
            size -= 1
        }
    }

    fun top() = head?.value
    fun isEmpty() = size == 0
    fun size() = size
}